File: delivery/api/src/main/java/org/example/delivery/api/domain/token/helper/JwtTokenHelper.java
package org.example.delivery.api.domain.token.helper;

import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import org.example.delivery.api.common.error.TokenErrorCode;
import org.example.delivery.api.common.exception.ApiException;
import org.example.delivery.api.domain.token.interfaces.TokenHelper;
import org.example.delivery.api.domain.token.model.TokenDto;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Component
public class JwtTokenHelper implements TokenHelper {

    @Value("${token.secret.key}")
    private String secretKey;

    @Value("${token.access-token.expired-hour}")
    private Long accessTokenExpiredHour;

    @Value("${token.refresh-token.expired-hour}")
    private Long refreshTokenExpiredHour;

    @Override
    public TokenDto issueAccessToken(Map<String, Object> data) {
        LocalDateTime expired = LocalDateTime.now().plusHours(accessTokenExpiredHour);
        Date expiredAt = Date.from(expired.atZone(ZoneId.systemDefault()).toInstant());

        SecretKey key = Keys.hmacShaKeyFor(secretKey.getBytes());

        String jwt = Jwts.builder()
                .signWith(key, SignatureAlgorithm.HS256)
                .setClaims(data)
                .setExpiration(expiredAt)
                .compact();

        return TokenDto.builder()
                .token(jwt)
                .expiredAt(expired)
                .build();
    }

    @Override
    public TokenDto issueRefreshToken(Map<String, Object> data) {
        LocalDateTime expired = LocalDateTime.now().plusHours(refreshTokenExpiredHour);
        Date expiredAt = Date.from(expired.atZone(ZoneId.systemDefault()).toInstant());

        SecretKey key = Keys.hmacShaKeyFor(secretKey.getBytes());

        String jwt = Jwts.builder()
                .signWith(key, SignatureAlgorithm.HS256)
                .setClaims(data)
                .setExpiration(expiredAt)
                .compact();

        return TokenDto.builder()
                .token(jwt)
                .expiredAt(expired)
                .build();
    }

    @Override
    public Map<String, Object> validationTokenWithThrow(String token) {
        SecretKey key = Keys.hmacShaKeyFor(secretKey.getBytes());

        JwtParser parser = Jwts.parser().setSigningKey(key).build();

        try {
            Jws<Claims> result = parser.parseClaimsJws(token);
            return new HashMap<>(result.getBody());
        } catch (Exception e) {
            if(e instanceof SignatureException){
                // 토큰이 유효하지 않을때
                throw new ApiException(TokenErrorCode.INVALID_TOKEN, e);
            }
            else if(e instanceof ExpiredJwtException){
                //  만료된 토큰
                throw new ApiException(TokenErrorCode.EXPIRED_TOKEN, e);
            }
            else{
                // 그외 에러
                throw new ApiException(TokenErrorCode.TOKEN_EXCEPTION, e);
            }
        }
    }
}

File: delivery/account/src/main/kotlin/org/example/delivery/account/domain/token/helper/JwtTokenHelper.kt
package org.example.delivery.account.domain.token.helper

import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.SignatureException
import org.example.delivery.account.domain.token.interfaces.TokenHelper
import org.example.delivery.account.domain.token.model.TokenDto
import org.example.delivery.common.error.TokenErrorCode
import org.example.delivery.common.exception.ApiException
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*
import kotlin.collections.HashMap

@Component
class JwtTokenHelper : TokenHelper {

    @Value("\${token.secret.key}")
    private val secretKey: String? = null

    @Value("\${token.access-token.expired-hour}")
    private val accessTokenExpiredHour: Long = 1

    @Value("\${token.refresh-token.expired-hour}")
    private val refreshTokenExpiredHour: Long = 12

    override fun issueAccessToken(data: Map<String, Any>): TokenDto? {
        val expired = LocalDateTime.now().plusDays(accessTokenExpiredHour)
        val expiredAt = expired.let {
            Date.from(it.atZone(ZoneId.systemDefault()).toInstant())
        }

        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val jwt = Jwts.builder()
            .signWith(key, Jwts.SIG.HS256)
            .claims(data)
            .expiration(expiredAt)
            .compact();

        return TokenDto(
            token = jwt,
            expiredAt = expired
        )
    }

    override fun issueRefreshToken(data: Map<String, Any>): TokenDto? {
        val expired = LocalDateTime.now().plusDays(refreshTokenExpiredHour)
        val expiredAt = expired.let {
            Date.from(it.atZone(ZoneId.systemDefault()).toInstant())
        }

        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val jwt = Jwts.builder()
            .signWith(key, Jwts.SIG.HS256)
            .claims(data)
            .expiration(expiredAt)
            .compact();

        return TokenDto(
            token = jwt,
            expiredAt = expired
        )
    }

    override fun validationTokenWithThrow(token: String?): Map<String, Any>? {
        val key = Keys.hmacShaKeyFor(secretKey?.toByteArray())

        val parser = Jwts.parser().verifyWith(key).build()

        return try {
            val result = token?.let {  parser.parseSignedClaims(it) }
            HashMap(result?.payload)
        } catch (e: Exception) {
            when(e) {
                is SignatureException -> {
                    throw ApiException(TokenErrorCode.INVALID_TOKEN, e)
                }
                is ExpiredJwtException -> {
                    throw ApiException(TokenErrorCode.EXPIRED_TOKEN, e)
                }
                else -> {
                    throw ApiException(TokenErrorCode.TOKEN_EXCEPTION, e)
                }
            }
        }
    }
}